Update description of aad and info. (#54)
cose-wg/HPKE: draft-ietf-cose-hpke.md

Examples of the COSE_Key for COSE-HPKE are shown in {{key-representation-example}}.

## Binding Keying Material to the Context of the COSE-HPKE Transaction {#binding-context}

The keying material derived in HPKE is originally bound to a specific transaction
between a specific sender and recipient, in a manner compliant with {{SP800-56A}}.

> Speficically, the KEM and KDF steps in HPKE supply sufficient context information

Reviewer: ```suggestion
Specifically, the KEM and KDF steps in HPKE supply sufficient context information
```

> Specifically, the KEM and KDF steps in HPKE supply sufficient context information

Reviewer 2: This is not correct.

The public keys are, of course, included in the computation. This is not what the context is about. The context is about the endpoint identifiers.

Author: Indeed, it's not accurate. I think your point is valid. 

Instead of claiming that information equivalent to COSE_KDF_Context is provided, it might be more appropriate to correctly refer SP800-53A and demonstrate that we have gained resistance to unknown key-share attacks, etc.

including the recipient and sender public keys, the string "HPKE-v1", the HPKE mode,
the specified ciphersuite represented by the combination of KEM, KDF and AEAD
algorithm IDs, and the usage label to the key derivation process.

> Therefore, there is no need to use the COSE_KDF_Context structure defined in {{RFC9053}}.

Reviewer 2: I disagree with this conclusion. If this is the case then why are there attacks like this https://datatracker.ietf.org/doc/rfc8844/ ?

I understand that it is annoying to provide this context information but unless we can proof that there are no problems we should be following the advice by cryptographers.

Author: Indeed, it's not accurate. I think your point is valid. 

Instead of claiming that information equivalent to COSE_KDF_Context is provided, it might be more appropriate to correctly refer SP800-53A and demonstrate that we have gained resistance to unknown key-share attacks, etc.

Author: I'd like you to read the following message I posted previously to COSE WG mailinglist.

https://mailarchive.ietf.org/arch/msg/cose/XjOlXbTvMRqxhdXhovv5K3fYNnw/

> First of all, COSE_KDF_Context is designed to be compliant with NIST SP800
> 56A#Section 5.8.2.1, but it's for *One-step Key Derivation*.
> 
> On the other hand, HPKE adopts Two-step Key Derivation internally (See
> RFC9180#KeySchedule<Role>(), LabeledExtract(), LabeledExpand()
> definitions), so we should refer not Section 5.8.2.1 but Section
> 5.8.2.2, *Two-step Key
> Derivation (Extraction-then-Expansion)*.
> ...

For the sake of improving interoperability and avoiding redundant context bindings
as much as possible, the COSE_KDF_Context MUST NOT be used with the COSE-HPKE algorithms.

## Auxiliary Authenticated Application Information {#aad}

HPKE has two options for specifying auxiliary authenticated information externally:
"info" and "aad".

In COSE-HPKE, since an HPKE encryption context is used per message, applications using
COSE-HPKE need only use either "info" or "aad" for specifying auxiliary authenticated
information.

Since COSE originally provides a way to supply the AAD (Additional Authenticated Data)
to the AEAD step, including an externaly supplied value (external_aad), in COSE-HPKE,

> using this AAD as "aad" for HPKE is more compatible and natural with the COSE convension.

Reviewer: ```suggestion
using this AAD as "aad" for HPKE is more compatible and natural with the COSE convention.
```

> Therefore, the "aad" for HPKE SHOULD be specified with the AAD that includes
> the external_aad, while the "info" should be specified with an empty string.

Reviewer: ```suggestion

In both "HPKE direct encryption" and "HPKE key encryption", the aad provided to the Seal and Open operations MUST be the protected headers.

When using "HPKE direct encryption":

This mode MUST be used with COSE_Encrypt0 (tag 16).
This mode MUST NOT be used with COSE_Encrypt (tag 96).

The `aad` provided to HPKE Seal and HPKE Open MUST be the protected bytes, as defined in {{Section 3 of RFC9052 }}.

This mode MAY be used with external aad, as described in {{Section 5.3 of RFC9052}}.

This mode MAY be used with COSE_KDF_Context as described in {{Section 5.2 of RFC9053}}.

Parameters to COSE_KDF_Context are known to both sender and receiver, and MAY be found in either the protected or unprotected headers of the COSE_Encrypt0 structure, or supplied directly to the encrypt or decrypt operations for this mode.

When using  "HPKE key encryption", 

This mode MUST NOT be used with COSE_Encrypt0 (tag 16).
This mode MUST be used with COSE_Encrypt (tag 96).

The `aad` provided to HPKE Seal and HPKE Open MUST be the protected bytes of the COSE_Recipient, as defined in {{Section 5.1 of RFC9052 }}.

This mode MAY be used with external aad, as described in {{Section 5.3 of RFC9052}}.

This mode MAY be used with COSE_KDF_Context as described in {{Section 5.2 of RFC9053}}.

Parameters to COSE_KDF_Context are known to both sender and receiver, and MAY be found in either the protected or unprotected headers of the COSE_Encrypt structure, the protected or unprotected headers of the COSE_Recipient structure, or supplied directly to the encrypt or decrypt operations for this mode.

```

Reviewer: I did implement support for both external aad:

https://github.com/transmute-industries/cose/pull/14/files#diff-edd54867e5a0ce9a217f529034c273c9391d56d9f1341190d4301c825462214f

and cose kdf context

https://github.com/transmute-industries/cose/pull/14/files#diff-19affcbbf817a6560c7519b299387e2e4c3bdc4440882f2650846c46ea4abd38

Reviewer 3: Yes, getting rid COSE_KDF_Context might be a good idea. At least worth exploring. One reason I dislike it a lot is because you have to go read obscure NIST documents to really understand what to put in it. That is way above the skill of the average engineer. We should do that work for them.

I think we want to be sure that what ever we propose can do all the important things that COSE_KDF_Context does. Those things are the addressing of threats. We might do this by defining a few more COSE headers that take the place of fields in COSE_KDF_Context.

I think there's a big difference between single-recipient and multiple-recipient.

Single-recipient is easy because we can just rest on the security of HPKE. I think the HPKE security design was more thorough than that of COSE. I would be content assuming that all the hashing and mixing inside COSE do everything that COSE_KDF_Context does.

Multiple-recipient seems more difficult because we're now outside of the HPKE design. I would like to see analysis to be sure what ever is proposed to replace COSE_KDF_Context does the job.

I believe this discussion is highly relative: https://mailarchive.ietf.org/arch/msg/cose/wrjgLbq42PRSho2pGXQhB4BXRdI/

I would expect an Enc_Structure to be used with multiple-recipient. That doesn't seem to be described. It will protect the COSE_Recipient headers.

Once you have that you can define new headers that replace the essential parts of COSE_KDF_Context.

I think we want to know that whatever we do defends against the attack presented at lamps at IETF 118. I think this might be addressable with a new header called "next_alg" or such.

Author: Thank you all for reviewing this PR.

> I am trying to find the paragraph in the HPKE RFC that says that.

https://www.rfc-editor.org/rfc/rfc9180.html#name-auxiliary-authenticated-app

Section 8.1 in RFC9180 says that:

> Applications that only use the single-shot APIs described in [Section 6](https://www.rfc-editor.org/rfc/rfc9180.html#single-shot-apis) should use the Setup info parameter for specifying auxiliary authenticated information. Implementations which only expose single-shot APIs should not allow applications to use both Setup info and Context aad or exporter_context auxiliary information parameters.

The COSE-HPKE does not necessarily have to use the Single-Shot API, but internally the situation is essentially the same as if the Single-Shot API were used, as an Encryption Context is created every time a message is generated. This section says we should use one or the other, so I thought it would be simple to follow it.

Specifically, the KEM and KDF steps in HPKE supply sufficient context information
including the recipient and sender public keys, the string "HPKE-v1", the HPKE mode,
the specified ciphersuite represented by the combination of KEM, KDF and AEAD
algorithm IDs, and the usage label to the key derivation process.

Therefore, there is no need to use the COSE_KDF_Context structure defined in {{RFC9053}}.
For the sake of improving interoperability and avoiding redundant context bindings
as much as possible, the COSE_KDF_Context MUST NOT be used with the COSE-HPKE algorithms.

## Auxiliary Authenticated Application Information {#aad}

HPKE has two options for specifying auxiliary authenticated information externally:
"info" and "aad".

In COSE-HPKE, since an HPKE encryption context is used per message, applications using
COSE-HPKE need only use either "info" or "aad" for specifying auxiliary authenticated
information.

Since COSE originally provides a way to supply the AAD (Additional Authenticated Data)
to the AEAD step, including an externaly supplied value (external_aad), in COSE-HPKE,
using this AAD as "aad" for HPKE is more compatible and natural with the COSE convention.

> In both "HPKE direct encryption" and "HPKE key encryption", the aad provided to the Seal and Open operations MUST be the protected headers.

Reviewer: ```suggestion
In both "HPKE direct encryption" and "HPKE key encryption", the aad provided to the Seal and Open operations MUST be the HPKE_Recipient_Enc_Structure.

HPKE_Recipient_Enc_Structure = CBOR ([
"HPKE_1_layer", layer1ProtectedHeaderBytes,
external_hpke_aad || empty buffer
])

```

Reviewer 3: Pretty sure this should align with COSE and be an Enc_Structure at the layer 1.

Reviewer: That's probably true.

When using "HPKE direct encryption":

This mode MUST be used with COSE_Encrypt0 (tag 16).
This mode MUST NOT be used with COSE_Encrypt (tag 96).

The `aad` provided to HPKE Seal and HPKE Open MUST be the protected bytes, as defined in {{Section 3 of RFC9052 }}.

This mode MAY be used with external aad, as described in {{Section 5.3 of RFC9052}}.

> This mode MAY be used with COSE_KDF_Context as described in {{Section 5.2 of RFC9053}}.

Reviewer: ```suggestion
```

> Parameters to COSE_KDF_Context are known to both sender and receiver, and MAY be found in either the protected or unprotected headers of the COSE_Encrypt0 structure, or supplied directly to the encrypt or decrypt operations for this mode.

Reviewer: ```suggestion
Parameters to COSE_KDF_Context MAY be found in either the protected or unprotected headers.

Unless understood in some consistent way by sender and receiver, they are ignored.
```

When using  "HPKE key encryption", 

This mode MUST NOT be used with COSE_Encrypt0 (tag 16).

> This mode MUST be used with COSE_Encrypt (tag 96).

Reviewer: ```suggestion
This mode MAY be used with COSE_Encrypt (tag 96).
```

Reviewer: Iirc tags are optional

> The `aad` provided to HPKE Seal and HPKE Open MUST be the protected bytes of the COSE_Recipient, as defined in {{Section 5.1 of RFC9052 }}.

Reviewer: ```suggestion
The `aad` provided to HPKE Seal and HPKE Open MUST be the HPKE_Recipient_Enc_Structure.

HPKE_Recipient_Enc_Structure = CBOR ([
"HPKE_2_layer", layer1ProtectedHeaderBytes, layer2RecipientProtectedHeaderBytes,
external_hpke_aad || empty buffer
])
```

Reviewer 3: I strongly favor just using the algorithm ID from layer1, rather than all the headers. If it's all the headers, you have to pass arbitrarily large amounts of data between layers, rather than just an algorithm ID. (Specifically, passing the algorithm ID was easy in t_cose. Passing all the headers would require restructuring and forcing a size limit on headers that is not there now).

I would like to not make this specific to HPKE. I think it should be used for the COSE -29 replacement too. So name it differently and use a different context string.

This is for a COSE_Recipient, so there's no external data as there is in layer 0 or in an Enc_Structure.

I think there should be optional private context data to replace the private data in COSE_KDF_Context that we are eliminating. It should be a bstr, empty if there is none.

Reviewer: I don't understand those recipient protected header is secured in the case that algorithm is the only parameter passed from recipient protected header... And layer 1 protected header is already variable length.

Treating both protected headers as bytes, seems simplest and safest.

Reviewer: Ilari did make a comment that protected headers don't always have alg.

That is a good argument to include alg, but I would still want to make sure that headers were committed too.

I think a spanning set of cases and a new enc structure would help land this.

Author: Hmm, I agree that a simple alternative to COSE_KDF_Context is necessary as a measure against the lamps attack, but in COSE-HPKE, wouldn't it be better to prohibit the use of non-AEAD algorithms for key wrapping (such as -29)?

This mode MAY be used with external aad, as described in {{Section 5.3 of RFC9052}}.

> This mode MAY be used with COSE_KDF_Context as described in {{Section 5.2 of RFC9053}}.

Reviewer: ```suggestion
```

> Parameters to COSE_KDF_Context are known to both sender and receiver, and MAY be found in either the protected or unprotected headers of the COSE_Encrypt structure, the protected or unprotected headers of the COSE_Recipient structure, or supplied directly to the encrypt or decrypt operations for this mode.

Reviewer: ```suggestion
Parameters related to COSE_KDF_Context MAY be found in either the protected or unprotected headers of the COSE_Encrypt structure, the protected or unprotected headers of the COSE_Recipient structure.

Unless understood in some consistent way by send are receiver, they are ignore.
```

to the AEAD step, including an externaly supplied value (external_aad), in COSE-HPKE,
using this AAD as "aad" for HPKE is more compatible and natural with the COSE convention.

In both "HPKE direct encryption" and "HPKE key encryption", the aad provided to the Seal and Open operations MUST be the protected headers.

> 

Reviewer 3: Hoping aad input becomes a newly defined Recipient_Structure as discussed on mailing list.

When using "HPKE direct encryption":

This mode MUST be used with COSE_Encrypt0 (tag 16).
This mode MUST NOT be used with COSE_Encrypt (tag 96).

The `aad` provided to HPKE Seal and HPKE Open MUST be the protected bytes, as defined in {{Section 3 of RFC9052 }}.

This mode MAY be used with external aad, as described in {{Section 5.3 of RFC9052}}.

> This mode MAY be used with COSE_KDF_Context as described in {{Section 5.2 of RFC9053}}.

Reviewer 3: Hoping we entirely get rid of COSE_KDF_Context as discussed on the mailing list.
